Enable mypy checks for entire evm module

cburgdorf opened the pull request:
### What was wrong?

So far, the `evm` modules was only partially checked with mypy since there were still many errors when running `mypy` over the entire `evm` module.

### How was it fixed?

This PR adds type hints only **selectively** for all the cases that are currently preventing us to run `mypy` on the entire `evm` module. It does **not add** all missing type annotations per file yet.

This in contrast to my previous PR which tried to add (more or less) all type hints per subdirectory in `evm`.

I will continue to send subsequent PRs to eventually have the entire code base typed but I believe it is a big when to be able to activate the `mypy` check for the entire `evm` module so that we can ensure new code already conforms to these rules.

#### Cute Animal Picture

![put a cute animal picture link inside the parentheses](https://images.unsplash.com/photo-1496316778157-e826f5bb47f2?ixlib=rb-0.3.5&ixid=eyJhcHBfaWQiOjEyMDd9&s=f221991eeab09a9b12c98afdce2f1199&auto=format&fit=crop&w=1355&q=80)

cburgdorf commented:
@pipermerriam @gsalgado This should be ready for review

gsalgado commented on evm/utils/datatypes.py:
```diff
@@ -55,6 +55,11 @@ def _get_top_level_keys(overrides: Dict[str, Any]) -> Iterator[str]:
+# https://github.com/python/mypy/wiki/Unsupported-Python-Features
+# Most of the cases where we silence mypy boil down to cases where
+# dynamic subclasses where generated through this method
+
```
I'm not sure I like the idea of having a blanket comment like this; I'd prefer if we had a separate, specific, comment for every place we need to silence mypy

cburgdorf replied:
I've put a short comment on the places where we had to add a `ignore` because of this and then referenced to here to lookup this longer explanation. I thought that's a good trade off so that I don't have to put a lengthy explanation on each `ignore`

gsalgado commented on evm/rlp/collations.py:
```diff
@@ -14,6 +14,8 @@
 from .transactions import BaseShardingTransaction
 from .headers import CollationHeader
 
+from typing import Type
```
This should be at the top of the file together with other stdlib imports

gsalgado commented on evm/rlp/blocks.py:
```diff
@@ -14,6 +14,8 @@
 from .transactions import BaseTransaction
 from .headers import BlockHeader
 
+from typing import Type  # noqa: F401
```
This should be at the top of the file together with other stdlib imports

gsalgado commented on evm/logic/logging.py:
```diff
@@ -4,6 +4,8 @@
 
 from evm import constants
 
+from typing import List  # noqa: F401
```
This should be at the top of the file together with other stdlib imports

gsalgado commented on evm/db/journal.py:
```diff
@@ -7,6 +7,8 @@
 from evm.db.backends.base import BaseDB
 from evm.exceptions import ValidationError
 
+from typing import Dict  # noqa: F401
```
This should be at the top of the file together with other stdlib imports

gsalgado commented on evm/computation.py:
```diff
@@ -48,6 +48,18 @@
+    Opcode
+)
+
+from typing import (  # noqa: F401
```
This should be at the top of the file together with other stdlib imports

gsalgado commented on evm/chains/tester/__init__.py:
```diff
@@ -19,33 +19,46 @@
+    BlockHeader
+)
+
+from typing import (
```
This should be at the top of the file together with other stdlib imports

cburgdorf replied:
Alright, will move all of them to the top

gsalgado commented on evm/__init__.py:
```diff
@@ -21,8 +21,8 @@
-logging.TRACE = TRACE_LEVEL_NUM
-logging.Logger.trace = trace
+logging.TRACE = TRACE_LEVEL_NUM  # type: ignore
+logging.Logger.trace = trace  # type: ignore
```
Why do we have to ignore those?

cburgdorf replied:
Because these properties do not exist and we are making them up out of thin air. I'll put a comment on it.

gsalgado commented on evm/vm/code_stream.py:
```diff
@@ -7,43 +7,48 @@
 
+from typing import (  # noqa: F401
+    Iterator,
+    Set
```
This should be at the top of the file together with other stdlib imports

gsalgado commented on evm/vm/forks/frontier/blocks.py:
```diff
@@ -28,6 +28,10 @@
 
+from typing import (  # noqa: F401
+    List
+)
```
This should be at the top of the file together with other stdlib imports

gsalgado commented on evm/vm/forks/frontier/vm_state.py:
```diff
@@ -40,9 +40,14 @@
+)
 from .validation import validate_frontier_transaction
 
+from typing import Type  # noqa: F401
```
This should be at the top of the file together with other stdlib imports

gsalgado commented on evm/vm/forks/sharding/vmc_handler.py:
```diff
@@ -32,6 +32,10 @@
 
+from typing import (  # noqa: F401
+    Dict
+)
```
This should be at the top of the file together with other stdlib imports

gsalgado commented on evm/vm_state.py:
```diff
@@ -22,6 +22,19 @@
+
+from typing import (  # noqa: F401
+    Type
+)
```
This should be at the top of the file together with other stdlib imports

cburgdorf commented:
@gsalgado Thank you for the review. I moved all the typing imports to the top and added a comment at the place where we are making up those two properties on the logging module.

The other `ignore` places already had comments attached and I also left the longer explainer comment on the `Configurable` class. Is that alright?

gsalgado commented on evm/vm/forks/sharding/vmc_handler.py:
```diff
@@ -1,3 +1,6 @@
+from typing import (  # noqa: F401
+    Dict
+)
```
There should be no blank lines between stdlib imports, and they should be ordered alphabetically

gsalgado commented on evm/vm_state.py:
```diff
@@ -1,3 +1,7 @@
+from typing import (  # noqa: F401
+    Type
+)
+
```
There should be no blank lines between stdlib imports, and they should be ordered alphabetically

gsalgado commented on evm/vm/code_stream.py:
```diff
@@ -1,3 +1,7 @@
+from typing import (  # noqa: F401
+    Iterator,
+    Set
+)
```
There should be no blank lines between stdlib imports, and they should be ordered alphabetically

gsalgado commented on evm/tools/test_builder/test_builder.py:
```diff
@@ -1,3 +1,8 @@
+    Dict,
+    List,
+    Any
+)
```
There should be no blank lines between stdlib imports, and they should be ordered alphabetically

gsalgado commented on evm/rlp/blocks.py:
```diff
@@ -1,3 +1,4 @@
+from typing import Type  # noqa: F401
```
There should be no blank lines between stdlib imports, and they should be ordered alphabetically

gsalgado commented on evm/logic/logging.py:
```diff
@@ -1,5 +1,5 @@
 from __future__ import absolute_import
-
+from typing import List  # noqa: F401
```
There should be no blank lines between stdlib imports, and they should be ordered alphabetically

gsalgado commented on evm/db/journal.py:
```diff
@@ -1,3 +1,5 @@
+from typing import Dict  # noqa: F401
```
There should be no blank lines between stdlib imports, and they should be ordered alphabetically

pipermerriam replied:
@carver implemented automated import sorting using [`isort`](https://github.com/timothycrosley/isort).  That's like a nice half-step towards what `black` does.

carver replied:
Yup, I've been happy with `isort`. Here are the configs we put in `tox.ini`, which should be close enough to the py-evm standards to drop in place with only a few edits:

https://github.com/ethereum/web3.py/blob/307dc879f18bec9cc0071fc2a6205c1d54b985f5/tox.ini#L9-L19

```
[isort]
combine_as_imports=True
force_sort_within_sections=True
include_trailing_comma=True
known_standard_library=pytest
known_third_party=lru,eth_tester  # customize
known_first_party=web3,ens  # customize
line_length=21
multi_line_output=3
skip=web3/main.py,web3/utils/windows.py  # customize
use_parentheses=True
```

cburgdorf replied:
Thanks for sharing @carver! I could give that a try. We also have a [tracking issue](https://github.com/ethereum/py-evm/issues/476) regarding black. I think moving to black would be my favorite option but it would be great to get more opinions on it :)

gsalgado commented on evm/computation.py:
```diff
@@ -7,6 +7,14 @@
     abstractmethod
 )
 
+from typing import (  # noqa: F401
```
There should be no blank lines between stdlib imports, and they should be ordered alphabetically

cburgdorf commented:
@gsalgado updated! Now this makes me even more supportive of switching to `black` ;)
